builtins/amp-video.js

```diff
@@ -36,7 +36,31 @@ export function installVideo(win) {
     }
 
     /** @override */
+    prerenderAllowed() {
+      return true;
+    }
+
+    /** @override */
+    buildCallback() {
+      /** @private */
+      this.realPreloadAttrValue_ = this.element.getAttribute('preload');
+      const posterAttr = this.element.getAttribute('poster');
+
+      if (!posterAttr) {
+        console/*OK*/.warn(
+            'No "poster" attribute has been provided for amp-video.');
+      }
+    }
+
+    /** @override */
     layoutCallback() {
+      const docState = documentStateFor(this.getWin());
+      // Disable video preload in prerender mode and reset it when rendered.
+      if (docState.getVisibilityState() === 'prerender') {
```

Comment:
I see. So basically the idea is to allow elements to listen to document-state and/or viewer state and react to event changes. For example in our case it would be something like:

```js
Viewer.addEventListener('state-change', event => {
  if (event.target.isVisible()) {
    this.resetPreloadAttr_();
  }
}.bind(this));
```